File: src/LTAdmin/Services/Business/GradeService.cs
using LTAdmin.Core;
using LTAdmin.Data;
using LTAdmin.Models.Dto;
using LTAdmin.Models.Entities;
using LTAdmin.Repositories;
using LTAdmin.Services.Common;
using LTAdmin.Services.Logging;
using LTAdmin.Services.Referentiel;
using LTAdmin.Services.Validation;

namespace LTAdmin.Services.Business;

/// <summary>
/// Saisie et suivi des notes de contrôle continu.
/// Règles : note entre 0 et barème (sauf absent), période non clôturée,
/// étudiant inscrit dans la classe de l'évaluation, unicité (évaluation, inscrit).
/// </summary>
public sealed class GradeService : ServiceBase
{
    private readonly EvaluationRepository _evaluations;
    private readonly EnrollmentRepository _inscriptions;

    public GradeService(AccessDatabase database, AppLogger logger, JournalService journal, ParametreService parametres)
        : base(database, logger, journal, parametres)
    {
        _evaluations = new EvaluationRepository(database);
        _inscriptions = new EnrollmentRepository(database);
    }

    public List<NoteSaisieRow> GetGrilleSaisie(int idEvaluation)
    {
        try
        {
            var evaluation = _evaluations.GetEvaluationDetail(idEvaluation);
            if (evaluation?.IdClasse is null) return new List<NoteSaisieRow>();
            return _evaluations.ListGrilleSaisie(idEvaluation, evaluation.IdClasse.Value);
        }
        catch (Exception ex)
        {
            Logger.Error("Chargement de la grille de saisie", ex);
            return new List<NoteSaisieRow>();
        }
    }

    /// <summary>Enregistre (insertion ou mise à jour) la note d'un inscrit.</summary>
    public Result UpsertNote(int idEvaluation, int idInscription, double? valeur, bool absent, string? observation, string codeUtr)
    {
        try
        {
            var evaluation = _evaluations.GetEvaluationDetail(idEvaluation);
            if (evaluation is null)
                return Result.Fail("Évaluation introuvable.", ErrorCodes.NotFound);
            if (evaluation.PeriodeCloturee == true)
                return Result.Fail("Période clôturée : les notes sont figées.", ErrorCodes.BusinessRule);
            if (string.IsNullOrWhiteSpace(observation)) observation = null;
            else if (observation.Length > 300)
                return Result.Fail("Observation : 300 caractères maximum.", ErrorCodes.Validation);

            var bareme = evaluation.Bareme ?? Parametres.BaremeDefaut;
            var validation = GradeValidator.ValidateNote(valeur, absent, bareme);
            if (!validation.IsValid) return Invalid(validation);

            var inscription = _inscriptions.GetById(idInscription);
            if (inscription is null)
                return Result.Fail("Inscription introuvable.", ErrorCodes.NotFound);
            if (inscription.IdClasse != evaluation.IdClasse)
                return Result.Fail("Cet étudiant n’est pas inscrit dans la classe de cette évaluation.", ErrorCodes.BusinessRule);

            var existing = _evaluations.GetNote(idEvaluation, idInscription);
            if (existing is null)
            {
                var id = _evaluations.InsertNote(new Note
                {
                    IdEvaluation = idEvaluation,
                    IdInscription = idInscription,
                    ValeurNote = absent ? null : valeur,
                    Absent = absent,
                    Observation = observation,
                    DateSaisie = DateTime.Now,
                    CodeUtr = codeUtr
                });
                Journal.LogCreation(codeUtr, Tables.Note, id, $"Éval {idEvaluation} / inscrit {idInscription}");
            }
            else
            {
                existing.ValeurNote = absent ? null : valeur;
                existing.Absent = absent;
                existing.Observation = observation;
                existing.DateSaisie = DateTime.Now;
                existing.CodeUtr = codeUtr;
                _evaluations.UpdateNote(existing);
                Journal.LogModification(codeUtr, Tables.Note, existing.IdNote, $"Éval {idEvaluation} / inscrit {idInscription}");
            }
            return Result.Ok("Note enregistrée.");
        }
        catch (Exception ex) { return Failure("Saisie d’une note", ex); }
    }

    public Result DeleteNote(int idNote, string codeUtr)
    {
        try
        {
            _evaluations.DeleteNote(idNote);
            Journal.LogSuppression(codeUtr, Tables.Note, idNote);
            return Result.Ok("Note supprimée.");
        }
        catch (Exception ex) { return Failure("Suppression d’une note", ex); }
    }

    public List<MoyenneMatiereRow> GetMoyennesMatiere(int? idClasse = null, int? idPeriode = null)
    {
        try { return _evaluations.ListMoyennesMatiere(idClasse, idPeriode); }
        catch (Exception ex)
        {
            Logger.Error("Lecture des moyennes par matière", ex);
            return new List<MoyenneMatiereRow>();
        }
    }

    public List<MoyennePeriodeRow> GetMoyennesPeriode(int? idClasse = null, int? idPeriode = null)
    {
        try { return _evaluations.ListMoyennesPeriode(idClasse, idPeriode); }
        catch (Exception ex)
        {
            Logger.Error("Lecture des moyennes par période", ex);
            return new List<MoyennePeriodeRow>();
        }
    }
}

File: ltadmin/services/business/grade_service.py
"""Saisie et suivi des notes de contrôle continu.

Règles : note entre 0 et barème (sauf absent), période non clôturée,
étudiant inscrit dans la classe de l'évaluation, unicité (évaluation, inscrit).
"""

from __future__ import annotations

import datetime as _dt
from typing import List, Optional

from ltadmin.core.result import Result
from ltadmin.data.access_database import AccessDatabase
from ltadmin.data.schema import Tables
from ltadmin.models.dto import MoyenneMatiereRow, MoyennePeriodeRow, NoteSaisieRow
from ltadmin.models.entities import Note
from ltadmin.repositories.enrollment_repository import EnrollmentRepository
from ltadmin.repositories.evaluation_repository import EvaluationRepository
from ltadmin.services.auth.guard import ensure_allowed
from ltadmin.services.auth.habilitations import Modules
from ltadmin.services.common import ServiceBase
from ltadmin.services.logging.app_logger import AppLogger
from ltadmin.services.logging.journal_service import JournalService
from ltadmin.services.referentiel.parametre_service import ParametreService
from ltadmin.services.validation.validation import GradeValidator


class GradeService(ServiceBase):

    def __init__(self, database: AccessDatabase, logger: AppLogger,
                 journal: JournalService, parametres: ParametreService):
        super().__init__(database, logger, journal, parametres)
        self._evaluations = EvaluationRepository(database)
        self._inscriptions = EnrollmentRepository(database)

    def get_grille_saisie(self, id_evaluation: int) -> List[NoteSaisieRow]:
        try:
            evaluation = self._evaluations.get_evaluation_detail(id_evaluation)
            if evaluation is None or evaluation.id_classe is None:
                return []
            return self._evaluations.list_grille_saisie(id_evaluation, evaluation.id_classe)
        except Exception as ex:
            self.logger.error("Chargement de la grille de saisie", ex)
            return []

    def upsert_note(self, id_evaluation: int, id_inscription: int,
                    valeur: Optional[float], absent: bool,
                    observation: Optional[str], code_utr: str) -> Result:
        """Enregistre (insertion ou mise à jour) la note d'un inscrit."""
        denied = ensure_allowed(code_utr, Modules.NOTES)
        if denied is not None:
            return denied
        try:
            evaluation = self._evaluations.get_evaluation_detail(id_evaluation)
            if evaluation is None:
                return Result.fail("Évaluation introuvable.", "INTROUVABLE")
            if evaluation.periode_cloturee is True:
                return Result.fail("Période clôturée : les notes sont figées.", "GESTION")
            if not observation or not observation.strip():
                observation = None
            elif len(observation) > 300:
                return Result.fail("Observation : 300 caractères maximum.", "VALIDATION")

            bareme = evaluation.bareme if evaluation.bareme is not None \
                else self.parametres.bareme_defaut
            validation = GradeValidator.validate_note(valeur, absent, bareme)
            if not validation.is_valid:
                return self.invalid(validation)

            inscription = self._inscriptions.get_by_id(id_inscription)
            if inscription is None:
                return Result.fail("Inscription introuvable.", "INTROUVABLE")
            if inscription.id_classe != evaluation.id_classe:
                return Result.fail(
                    "Cet étudiant n’est pas inscrit dans la classe de cette évaluation.",
                    "GESTION")

            existing = self._evaluations.get_note(id_evaluation, id_inscription)
            if existing is None:
                id_note = self._evaluations.insert_note(Note(
                    id_evaluation=id_evaluation,
                    id_inscription=id_inscription,
                    valeur_note=None if absent else valeur,
                    absent=absent,
                    observation=observation,
                    date_saisie=_dt.datetime.now(),
                    code_utr=code_utr))
                self.journal.log_creation(
                    code_utr, Tables.NOTE, id_note,
                    f"Éval {id_evaluation} / inscrit {id_inscription}")
            else:
                existing.valeur_note = None if absent else valeur
                existing.absent = absent
                existing.observation = observation
                existing.date_saisie = _dt.datetime.now()
                existing.code_utr = code_utr
                self._evaluations.update_note(existing)
                self.journal.log_modification(
                    code_utr, Tables.NOTE, existing.id_note,
                    f"Éval {id_evaluation} / inscrit {id_inscription}")
            return Result.ok("Note enregistrée.")
        except Exception as ex:
            return self.failure("Saisie d’une note", ex)

    def delete_note(self, id_note: int, code_utr: str) -> Result:
        denied = ensure_allowed(code_utr, Modules.NOTES)
        if denied is not None:
            return denied
        try:
            self._evaluations.delete_note(id_note)
            self.journal.log_suppression(code_utr, Tables.NOTE, id_note)
            return Result.ok("Note supprimée.")
        except Exception as ex:
            return self.failure("Suppression d’une note", ex)

    def get_moyennes_matiere(self, id_classe: Optional[int] = None,
                             id_periode: Optional[int] = None) -> List[MoyenneMatiereRow]:
        try:
            return self._evaluations.list_moyennes_matiere(id_classe, id_periode)
        except Exception as ex:
            self.logger.error("Lecture des moyennes par matière", ex)
            return []

    def get_moyennes_periode(self, id_classe: Optional[int] = None,
                             id_periode: Optional[int] = None) -> List[MoyennePeriodeRow]:
        try:
            return self._evaluations.list_moyennes_periode(id_classe, id_periode)
        except Exception as ex:
            self.logger.error("Lecture des moyennes par période", ex)
            return []
